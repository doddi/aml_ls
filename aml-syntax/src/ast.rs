use std::collections::HashMap;
use serde::Serialize;
use aml_core::Location;

#[derive(Debug, Default)]
pub struct Ast {
    pub nodes: Vec<AstNode>,
    pub variables: HashMap<String, Location>,
    pub scopes: Vec<Scope>,
}

#[derive(Debug)]
pub enum AstNode {
    String {
        value: Location,
    },
    Text {
        value: Option<Box<AstNode>>,
        attributes: Vec<AstNode>,
        children: Vec<AstNode>,
        location: Location,
    },
    Span {
        value: Option<Box<AstNode>>,
        attributes: Vec<AstNode>,
        location: Location,
    },
    Identifier {
        value: Location,
    },
    Attribute {
        name: Box<AstNode>,
        value: Box<Expr>,
    },
}

#[derive(Debug, Serialize)]
pub struct Scope {
    pub variables: Vec<String>,
    pub parent: Option<usize>,
}

#[derive(Debug, Clone, Serialize)]
pub enum Expr {
    Unary {
        op: aml_token::Operator,
        expr: Box<Expr>,
    },
    Binary {
        lhs: Box<Expr>,
        rhs: Box<Expr>,
        op: aml_token::Operator,
    },
    Ident {
        location: Location,
    },
    String {
        location: Location,
    },
    Call {
        fun: Box<Expr>,
        args: Vec<Expr>,
    },
    Primitive(aml_token::Primitive),
    ArrayIndex {
        lhs: Box<Expr>,
        index: Box<Expr>,
    },
    List(Vec<Expr>),
    Map {
        items: Vec<(Expr, Expr)>,
    },
}