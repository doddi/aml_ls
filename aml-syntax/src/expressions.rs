use aml_core::Location;
use aml_token::{Operator, TokenKind, Tokens};

use crate::ast::Expr;

pub mod precedences {
    pub const INITIAL: u8 = 0;
    pub const CONDITIONAL: u8 = 2;
    pub const EQUALITY: u8 = 3;
    pub const LOGICAL: u8 = 4;
    pub const SUM: u8 = 5;
    pub const PRODUCT: u8 = 6;
    pub const PREFIX: u8 = 8;
    pub const CALL: u8 = 10;
    pub const SUBCRIPT: u8 = 11;
}

fn get_precedence(op: Operator) -> u8 {
    match op {
        Operator::Dot | Operator::LBracket => precedences::SUBCRIPT,
        Operator::LParen => precedences::CALL,
        Operator::Mul | Operator::Div | Operator::Mod => precedences::PRODUCT,
        Operator::Plus | Operator::Minus => precedences::SUM,
        Operator::GreaterThan
        | Operator::GreaterThanOrEqual
        | Operator::LessThan
        | Operator::LessThanOrEqual => precedences::LOGICAL,
        Operator::EqualEqual | Operator::NotEqual => precedences::EQUALITY,
        Operator::Or | Operator::And | Operator::Either => precedences::CONDITIONAL,

        _ => precedences::INITIAL,
    }
}

pub fn parse_expression(tokens: &mut Tokens) -> Expr {
    parse_expression_inner(tokens, precedences::INITIAL)
}

fn parse_expression_inner(tokens: &mut Tokens, precedence: u8) -> Expr {
    let next = tokens.next_no_indent();
    let location = next.location();

    let mut lhs = match next.kind() {
        TokenKind::Operator(Operator::LBracket) => parse_collection(tokens, location),
        TokenKind::Operator(Operator::LCurly) => parse_map(tokens, location),
        TokenKind::Operator(Operator::LParen) => {
            let lhs = parse_expression_inner(tokens, precedences::INITIAL);
            assert!(matches!(
                tokens.next_no_indent().kind(),
                TokenKind::Operator(Operator::RParen)
            ));
            lhs
        }
        TokenKind::Operator(op @ Operator::Minus) => parse_unary_expression(tokens, op, location),
        TokenKind::Operator(op @ Operator::Not) => parse_unary_expression(tokens, op, location),
        TokenKind::String(location) => Expr::String { location },
        TokenKind::Primitive(primitive) => Expr::Primitive {
            value: primitive,
            location,
        },
        TokenKind::Identifier(location) => Expr::Ident { location },

        // all of these are invalid
        TokenKind::Operator(Operator::RBracket)
        | TokenKind::Operator(Operator::RParen)
        | TokenKind::Operator(Operator::Association)
        | TokenKind::Operator(Operator::RCurly)
        | TokenKind::Operator(Operator::Mul)
        | TokenKind::Operator(Operator::Div)
        | TokenKind::Operator(Operator::Mod)
        | TokenKind::Operator(Operator::PlusEqual)
        | TokenKind::Operator(Operator::MinusEqual)
        | TokenKind::Operator(Operator::MulEqual)
        | TokenKind::Operator(Operator::DivEqual)
        | TokenKind::Operator(Operator::ModEqual)
        | TokenKind::Operator(Operator::Colon)
        | TokenKind::Operator(Operator::Comma)
        | TokenKind::Operator(Operator::Plus)
        | TokenKind::Operator(Operator::Dot)
        | TokenKind::Operator(Operator::GreaterThan)
        | TokenKind::Operator(Operator::GreaterThanOrEqual)
        | TokenKind::Operator(Operator::LessThan)
        | TokenKind::Operator(Operator::LessThanOrEqual)
        | TokenKind::Operator(Operator::EqualEqual)
        | TokenKind::Operator(Operator::NotEqual)
        | TokenKind::Operator(Operator::Or)
        | TokenKind::Operator(Operator::And)
        | TokenKind::Operator(Operator::Either)
        | TokenKind::Equal
        | TokenKind::For
        | TokenKind::In
        | TokenKind::If
        | TokenKind::Else
        | TokenKind::Switch
        | TokenKind::Case
        | TokenKind::Default
        | TokenKind::With
        | TokenKind::As
        | TokenKind::Component
        | TokenKind::ComponentSlot
        | TokenKind::Decl
        | TokenKind::Local
        | TokenKind::Global
        | TokenKind::Eof
        | TokenKind::Error(_)
        | TokenKind::Element(_)
        | TokenKind::Container(_)
        | TokenKind::Newline => {
            return Expr::Error {
                location,
                token: next.kind(),
            };
        }

        TokenKind::Indent(_) => unreachable!(),
    };

    loop {
        let TokenKind::Operator(op) = tokens.peek_skip_indent().kind() else {
            return lhs;
        };

        match op {
            Operator::Association
            | Operator::PlusEqual
            | Operator::MinusEqual
            | Operator::MulEqual
            | Operator::DivEqual
            | Operator::ModEqual => {
                let token = tokens.next_no_indent();
                return Expr::Error {
                    location: token.location(),
                    token: token.kind(),
                };
            }
            _ => {}
        }

        let op_precedence = get_precedence(op);

        if precedence >= op_precedence {
            break;
        }

        tokens.consume();

        match op {
            Operator::LParen => {
                lhs = parse_function(tokens, lhs, location);
                continue;
            }
            Operator::LBracket => {
                let index = parse_expression_inner(tokens, precedences::INITIAL);
                let next_token = tokens.next_no_indent();

                assert!(matches!(
                    next_token.kind(),
                    TokenKind::Operator(Operator::RBracket)
                ));

                lhs = Expr::ArrayIndex {
                    lhs: Box::new(lhs),
                    location: location.merge(next_token.location()),
                    index: Box::new(index),
                };

                continue;
            }
            _ => {}
        }

        let rhs = parse_expression_inner(tokens, op_precedence);
        let location = location.merge(rhs.location());
        lhs = Expr::Binary {
            op,
            lhs: Box::new(lhs),
            rhs: Box::new(rhs),
            location,
        };
    }

    lhs
}

fn parse_unary_expression(tokens: &mut Tokens, operator: Operator, location: Location) -> Expr {
    let expr = parse_expression_inner(tokens, precedences::PREFIX);
    let location = location.merge(expr.location());

    Expr::Unary {
        op: operator,
        expr: Box::new(expr),
        location,
    }
}

fn parse_collection(tokens: &mut Tokens, start_location: Location) -> Expr {
    let mut items = vec![];

    let end_location = loop {
        let next_token = tokens.peek_skip_indent();
        match next_token.kind() {
            TokenKind::Newline => {
                tokens.consume();
                break next_token.location();
            }
            TokenKind::Eof => {
                items.push(Expr::Error {
                    location: next_token.location(),
                    token: next_token.kind(),
                });
                break next_token.location();
            }
            TokenKind::Operator(Operator::RBracket) => {
                tokens.consume();
                break next_token.location();
            }
            TokenKind::Operator(Operator::Comma) => {
                tokens.consume();
                continue;
            }
            _ => {}
        }
        items.push(parse_expression_inner(tokens, precedences::INITIAL));
    };

    let location = start_location.merge(end_location);
    Expr::List { items, location }
}

fn parse_map(tokens: &mut Tokens, start_location: Location) -> Expr {
    let mut items = vec![];

    let end_location = loop {
        let next_token = tokens.peek_skip_indent();
        match next_token.kind() {
            TokenKind::Newline => {
                tokens.consume();
                continue;
            }
            TokenKind::Operator(Operator::Comma) => {
                tokens.consume();
                continue;
            }
            TokenKind::Operator(Operator::RCurly) => {
                tokens.consume();
                break next_token.location();
            }
            _ => {}
        }

        let key = parse_expression_inner(tokens, precedences::INITIAL);

        match tokens.peek_skip_indent().kind() {
            TokenKind::Operator(Operator::Colon) => tokens.consume(),
            _ => {
                items.push((
                    key,
                    Expr::Error {
                        location: next_token.location(),
                        token: next_token.kind(),
                    },
                ));
                break next_token.location();
            }
        }
        let value = parse_expression_inner(tokens, precedences::INITIAL);
        items.push((key, value));
    };

    let location = start_location.merge(end_location);

    Expr::Map { items, location }
}

fn parse_function(tokens: &mut Tokens, lhs: Expr, start_location: Location) -> Expr {
    let mut args = vec![];

    let end_location = loop {
        let next_token = tokens.peek_skip_indent();
        match next_token.kind() {
            TokenKind::Operator(Operator::Comma) => {
                tokens.consume();
                continue;
            }
            TokenKind::Operator(Operator::RParen) => {
                tokens.consume();
                break next_token.location();
            }
            TokenKind::Eof => {
                args.push(Expr::Error {
                    location: next_token.location(),
                    token: next_token.kind(),
                });
                break next_token.location();
            }
            _ => {}
        }
        args.push(parse_expression_inner(tokens, precedences::INITIAL));
    };

    let location = start_location.merge(end_location);

    Expr::Call {
        fun: Box::new(lhs),
        args,
        location,
    }
}

#[cfg(test)]
pub(crate) mod test {
    use aml_core::Location;
    use aml_token::{Lexer, Primitive};
    use serde::Serialize;

    use super::*;

    #[derive(Debug, Serialize)]
    pub enum SnapshotExpr<'ast> {
        Unary {
            original: &'ast str,
            op: Operator,
            expr: Box<SnapshotExpr<'ast>>,
            location: Location,
        },
        Binary {
            original: &'ast str,
            lhs: Box<SnapshotExpr<'ast>>,
            rhs: Box<SnapshotExpr<'ast>>,
            op: Operator,
            location: Location,
        },
        Ident {
            value: &'ast str,
            location: Location,
        },
        String {
            value: &'ast str,
            location: Location,
        },
        Call {
            original: &'ast str,
            fun: Box<SnapshotExpr<'ast>>,
            args: Vec<SnapshotExpr<'ast>>,
            location: Location,
        },
        Primitive {
            original: &'ast str,
            value: Primitive,
            location: Location,
        },
        ArrayIndex {
            original: &'ast str,
            lhs: Box<SnapshotExpr<'ast>>,
            index: Box<SnapshotExpr<'ast>>,
            location: Location,
        },
        List {
            original: &'ast str,
            items: Vec<SnapshotExpr<'ast>>,
            location: Location,
        },
        Map {
            original: &'ast str,
            items: Vec<(SnapshotExpr<'ast>, SnapshotExpr<'ast>)>,
            location: Location,
        },
        Error {
            original: &'ast str,
            location: Location,
            token: TokenKind,
        },
    }

    impl<'ast> SnapshotExpr<'ast> {
        pub fn from_expr(expr: Expr, content: &'ast str) -> Self {
            match expr {
                Expr::Unary { op, expr, location } => Self::Unary {
                    op,
                    location,
                    original: &content[location.to_range()],
                    expr: Box::new(SnapshotExpr::from_expr(*expr, content)),
                },
                Expr::Binary {
                    lhs,
                    rhs,
                    op,
                    location,
                } => Self::Binary {
                    op,
                    location,
                    original: &content[location.to_range()],
                    lhs: Box::new(SnapshotExpr::from_expr(*lhs, content)),
                    rhs: Box::new(SnapshotExpr::from_expr(*rhs, content)),
                },
                Expr::Ident { location } => Self::Ident {
                    value: &content[location.to_range()],
                    location,
                },
                Expr::String { location } => Self::String {
                    value: &content[location.to_range()],
                    location,
                },
                Expr::Call {
                    fun,
                    args,
                    location,
                } => Self::Call {
                    fun: Box::new(SnapshotExpr::from_expr(*fun, content)),
                    args: args
                        .into_iter()
                        .map(|e| SnapshotExpr::from_expr(e, content))
                        .collect(),
                    location,
                    original: &content[location.to_range()],
                },
                Expr::Primitive { value, location } => Self::Primitive {
                    value,
                    location,
                    original: &content[location.to_range()],
                },
                Expr::ArrayIndex {
                    lhs,
                    index,
                    location,
                } => Self::ArrayIndex {
                    location,
                    original: &content[location.to_range()],
                    lhs: Box::new(SnapshotExpr::from_expr(*lhs, content)),
                    index: Box::new(SnapshotExpr::from_expr(*index, content)),
                },
                Expr::List { items, location } => Self::List {
                    location,
                    original: &content[location.to_range()],
                    items: items
                        .into_iter()
                        .map(|e| SnapshotExpr::from_expr(e, content))
                        .collect(),
                },
                Expr::Map { items, location } => Self::Map {
                    location,
                    original: &content[location.to_range()],
                    items: items
                        .into_iter()
                        .map(|(l, r)| {
                            (
                                SnapshotExpr::from_expr(l, content),
                                SnapshotExpr::from_expr(r, content),
                            )
                        })
                        .collect(),
                },
                Expr::Error { location, token } => Self::Error {
                    token,
                    location,
                    original: &content[location.to_range()],
                },
            }
        }
    }

    fn parse(input: &str) -> SnapshotExpr<'_> {
        let tokens = Lexer::new(input).collect();
        let mut tokens = Tokens::new(tokens, input.len());
        let expression = parse_expression_inner(&mut tokens, precedences::INITIAL);
        SnapshotExpr::from_expr(expression, input)
    }

    #[test]
    fn test_simple_add() {
        let input = "1 + 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_sub() {
        let input = "1 - 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_mul() {
        let input = "5 + 1 * 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_div() {
        let input = "5 - 1 / 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_expr_grouping_precedence() {
        let input = "(5 + 1) * 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_function_call() {
        let input = "fun(1, a + 2 * 3, 3)";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_function_no_args() {
        let input = "f()";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_array_index() {
        let input = "array[0][1]";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_map_lookup() {
        let input = "map['key']";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_dot_lookup() {
        let input = "a.b.c";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_modulo() {
        let input = "5 + 1 % 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_list() {
        let input = "[1, 2, a, 4]";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_nested_list() {
        let input = "[1, [2, 3, [4, 5]]]";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_equality() {
        let input = "1 == 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_map_declaration() {
        let input = "{a: 1, b: c}";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_and() {
        let input = "1 == 2 && 3 == 4";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_not() {
        let input = "1 != 2 && 3 != 4";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_either() {
        let input = "a ? b ? c";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_unclosed_parenthesis() {
        let input = "func(1, 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_unclosed_bracket() {
        let input = "[1, 2, 3";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_unclosed_curly_brace() {
        let input = "{a: 1, b: 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_missing_comma_in_list() {
        let input = "[1 + 12 * a b c - 13]";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_missing_comma_in_function_args() {
        let input = "foo(1 2 3)";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_missing_operand() {
        let input = "1 + ";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_missing_left_operand() {
        let input = "* 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    // TODO: this test panics right now due to an asssertion, but it should produce a nice error
    // node for error reporting
    // #[test]
    // fn test_mixed_brackets() {
    //     let input = "(1 + [2)";
    //     insta::assert_yaml_snapshot!(parse(input));
    // }

    #[test]
    fn test_incomplete_binary_expression() {
        let input = "a +";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_double_operators() {
        let input = "1 + + 2";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_empty_map_entry() {
        let input = "{: 1, b:}";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_invalid_map_key() {
        let input = "{1+2: 3}";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_incomplete_dot_lookup() {
        let input = "obj.";
        insta::assert_yaml_snapshot!(parse(input));
    }

    // TODO: this also hits an assertion error, which it shouldn't
    // #[test]
    // fn test_empty_array_index() {
    //     let input = "array[]";
    //     insta::assert_yaml_snapshot!(parse(input));
    // }

    #[test]
    fn test_invalid_either() {
        let input = "a ? ";
        insta::assert_yaml_snapshot!(parse(input));
    }

    #[test]
    fn test_consecutive_either() {
        let input = "a ? ? c";
        insta::assert_yaml_snapshot!(parse(input));
    }
}
