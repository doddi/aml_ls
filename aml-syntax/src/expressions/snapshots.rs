use aml_core::Location;
use aml_token::{Operator, Primitive, TokenKind};
use serde::Serialize;

use crate::ast::Expr;

#[derive(Debug, Serialize)]
pub enum SnapshotExpr<'ast> {
    Unary {
        original: &'ast str,
        op: Operator,
        expr: Box<SnapshotExpr<'ast>>,
        location: Location,
    },
    Binary {
        original: &'ast str,
        lhs: Box<SnapshotExpr<'ast>>,
        rhs: Box<SnapshotExpr<'ast>>,
        op: Operator,
        location: Location,
    },
    Ident {
        value: &'ast str,
        location: Location,
    },
    String {
        value: &'ast str,
        location: Location,
    },
    Call {
        original: &'ast str,
        fun: Box<SnapshotExpr<'ast>>,
        args: Vec<SnapshotExpr<'ast>>,
        location: Location,
    },
    Primitive {
        original: &'ast str,
        value: Primitive,
        location: Location,
    },
    ArrayIndex {
        original: &'ast str,
        lhs: Box<SnapshotExpr<'ast>>,
        index: Box<SnapshotExpr<'ast>>,
        location: Location,
    },
    List {
        original: &'ast str,
        items: Vec<SnapshotExpr<'ast>>,
        location: Location,
    },
    Map {
        original: &'ast str,
        items: Vec<(SnapshotExpr<'ast>, SnapshotExpr<'ast>)>,
        location: Location,
    },
    Error {
        original: &'ast str,
        location: Location,
        token: TokenKind,
    },
}

impl<'ast> SnapshotExpr<'ast> {
    pub fn from_expr(expr: Expr, content: &'ast str) -> Self {
        match expr {
            Expr::Unary { op, expr, location } => Self::Unary {
                op,
                location,
                original: &content[location.to_range()],
                expr: Box::new(SnapshotExpr::from_expr(*expr, content)),
            },
            Expr::Binary {
                lhs,
                rhs,
                op,
                location,
            } => Self::Binary {
                op,
                location,
                original: &content[location.to_range()],
                lhs: Box::new(SnapshotExpr::from_expr(*lhs, content)),
                rhs: Box::new(SnapshotExpr::from_expr(*rhs, content)),
            },
            Expr::Ident { location } => Self::Ident {
                value: &content[location.to_range()],
                location,
            },
            Expr::String { location } => Self::String {
                value: &content[location.to_range()],
                location,
            },
            Expr::Call {
                fun,
                args,
                location,
            } => Self::Call {
                fun: Box::new(SnapshotExpr::from_expr(*fun, content)),
                args: args
                    .into_iter()
                    .map(|e| SnapshotExpr::from_expr(e, content))
                    .collect(),
                location,
                original: &content[location.to_range()],
            },
            Expr::Primitive { value, location } => Self::Primitive {
                value,
                location,
                original: &content[location.to_range()],
            },
            Expr::ArrayIndex {
                lhs,
                index,
                location,
            } => Self::ArrayIndex {
                location,
                original: &content[location.to_range()],
                lhs: Box::new(SnapshotExpr::from_expr(*lhs, content)),
                index: Box::new(SnapshotExpr::from_expr(*index, content)),
            },
            Expr::List { items, location } => Self::List {
                location,
                original: &content[location.to_range()],
                items: items
                    .into_iter()
                    .map(|e| SnapshotExpr::from_expr(e, content))
                    .collect(),
            },
            Expr::Map { items, location } => Self::Map {
                location,
                original: &content[location.to_range()],
                items: items
                    .into_iter()
                    .map(|(l, r)| {
                        (
                            SnapshotExpr::from_expr(l, content),
                            SnapshotExpr::from_expr(r, content),
                        )
                    })
                    .collect(),
            },
            Expr::Error { location, token } => Self::Error {
                token,
                location,
                original: &content[location.to_range()],
            },
        }
    }
}
