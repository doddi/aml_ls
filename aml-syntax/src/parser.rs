use aml_token::{Element, Operator, TokenKind, Tokens};

use crate::ast::{Ast, AstNode, Scope};
use crate::expressions::parse_expression;

pub struct Parser {
    scope_stack: Vec<usize>,
    tokens: Tokens,
    ast: Ast,
}

impl Parser {
    pub fn new(mut tokens: Tokens) -> Self {
        tokens.consume_newlines();

        Self {
            tokens,
            ast: Ast::default(),
            scope_stack: Vec::new(),
        }
    }

    pub fn parse(mut self) -> Ast {
        let base_indent = match self.tokens.peek().kind() {
            TokenKind::Indent(indent) => indent,
            _ => 0,
        };

        self.ast.nodes = self.parse_block(base_indent);

        self.ast
    }

    fn parse_block(&mut self, block_indent: usize) -> Vec<AstNode> {
        self.add_scope();

        let mut nodes = vec![];

        loop {
            self.tokens.consume_newlines();

            let current_indent = match self.tokens.peek().kind() {
                TokenKind::Indent(i) => {
                    self.tokens.consume();
                    i
                }
                TokenKind::Eof => break,
                _ => 0,
            };

            if current_indent < block_indent {
                break;
            }

            if current_indent > block_indent {
                break;
            }

            if self.tokens.peek().kind() == TokenKind::Eof {
                break;
            }

            nodes.push(self.parse_node(current_indent));
        }

        self.pop_scope();
        nodes
    }

    fn parse_node(&mut self, current_indent: usize) -> AstNode {
        match self.tokens.peek_skip_indent().kind() {
            TokenKind::Element(element) => self.parse_element(element, current_indent),
            TokenKind::String(_) => self.parse_string(),
            TokenKind::Identifier(_) => self.parse_identifier(),
            TokenKind::Decl => self.parse_declaration(),
            t => todo!("unhandled token in parse_node: {t:?}"),
        }
    }

    fn parse_element(&mut self, element: Element, current_indent: usize) -> AstNode {
        match element {
            Element::Text => self.parse_text(current_indent),
            Element::Span => self.parse_span(),
            t => todo!("unhandled element: {t:?}"),
        }
    }

    fn parse_text(&mut self, current_indent: usize) -> AstNode {
        let text = self.tokens.next_token();
        let location = text.location();

        self.tokens.consume_all_whitespace();
        let attributes = self.parse_optional_attributes();

        self.tokens.consume_all_whitespace();
        let value = match self.tokens.peek().kind() {
            TokenKind::String(_) => Some(Box::new(self.parse_string())),
            _ => None,
        };

        self.tokens.consume_newlines();

        let next_indent = match self.tokens.peek().kind() {
            TokenKind::Indent(i) => i,
            _ => 0,
        };

        let children = match next_indent > current_indent {
            true => self.parse_block(next_indent),
            false => vec![],
        };

        AstNode::Text {
            value,
            attributes,
            children,
            location,
        }
    }

    fn parse_span(&mut self) -> AstNode {
        let span = self.tokens.next_token();
        let location = span.location();

        self.tokens.consume_all_whitespace();
        let attributes = self.parse_optional_attributes();

        self.tokens.consume_all_whitespace();
        let value = match self.tokens.peek().kind() {
            TokenKind::String(_) => Some(Box::new(self.parse_string())),
            _ => None,
        };

        AstNode::Span {
            value,
            attributes,
            location,
        }
    }

    fn parse_string(&mut self) -> AstNode {
        let token = self.tokens.next_token();
        let value = token.location();
        AstNode::String { value }
    }

    fn parse_identifier(&mut self) -> AstNode {
        let identifier = self.tokens.next_token();
        let location = identifier.location();
        AstNode::Identifier { value: location }
    }

    fn parse_declaration(&mut self) -> AstNode {
        let keyword = self.tokens.next_token();
        let location = keyword.location();

        self.tokens.consume_indent();
        let name = self.parse_identifier();

        self.tokens.consume_indent();
        self.tokens.consume(); // consume equal sign
        let value = parse_expression(&mut self.tokens);

        AstNode::Declaration {
            name: Box::new(name),
            value,
            location,
        }
    }

    fn parse_optional_attributes(&mut self) -> Vec<AstNode> {
        if self.tokens.peek_skip_indent().kind() == TokenKind::Operator(Operator::LBracket) {
            return self.parse_attributes();
        }

        vec![]
    }

    fn parse_attributes(&mut self) -> Vec<AstNode> {
        self.tokens.consume(); // consume LBracket

        let mut attributes = vec![];

        loop {
            match self.tokens.peek_skip_indent().kind() {
                TokenKind::Operator(Operator::RBracket) => {
                    self.tokens.consume();
                    break;
                }
                TokenKind::Newline => {
                    self.tokens.consume();
                    continue;
                }
                _ => {}
            }

            let name = self.parse_identifier();
            self.tokens.consume_all_whitespace();

            // TODO: this is a syntax error if there is no colon
            if self.tokens.peek().kind() == TokenKind::Operator(Operator::Colon) {
                self.tokens.consume();
            }

            self.tokens.consume_all_whitespace();
            let value = parse_expression(&mut self.tokens);

            attributes.push(AstNode::Attribute {
                name: Box::new(name),
                value,
            });

            self.skip_optional_comma();
        }

        attributes
    }

    fn skip_optional_comma(&mut self) {
        if self.tokens.peek_skip_indent().kind() == TokenKind::Operator(Operator::Comma) {
            self.tokens.consume();
        }
    }

    fn add_scope(&mut self) {
        let scope_id = self.ast.scopes.len();
        let parent = self.scope_stack.last().copied();

        self.ast.scopes.push(Scope {
            variables: Vec::new(),
            parent,
        });

        self.scope_stack.push(scope_id);
    }

    fn pop_scope(&mut self) {
        self.scope_stack.pop();
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use aml_core::Location;
    use aml_token::Lexer;
    use serde::Serialize;

    use super::*;
    use crate::expressions::test::SnapshotExpr;

    #[derive(Debug, Serialize)]
    struct SnapshotAst<'ast> {
        pub nodes: Vec<SnapshotAstNode<'ast>>,
        pub variables: HashMap<String, Location>,
        pub scopes: Vec<Scope>,
    }

    impl<'ast> SnapshotAst<'ast> {
        pub fn from_ast(ast: Ast, content: &'ast str) -> Self {
            Self {
                nodes: ast
                    .nodes
                    .into_iter()
                    .map(|n| SnapshotAstNode::from_node(n, content))
                    .collect(),
                variables: ast.variables,
                scopes: ast.scopes,
            }
        }
    }

    #[derive(Debug, Serialize)]
    enum SnapshotAstNode<'ast> {
        String {
            value: Location,
            text: &'ast str,
        },
        Text {
            value: Option<Box<SnapshotAstNode<'ast>>>,
            attributes: Vec<SnapshotAstNode<'ast>>,
            children: Vec<SnapshotAstNode<'ast>>,
            location: Location,
            text: &'ast str,
        },
        Span {
            value: Option<Box<SnapshotAstNode<'ast>>>,
            attributes: Vec<SnapshotAstNode<'ast>>,
            location: Location,
            text: &'ast str,
        },
        Identifier {
            value: Location,
            text: &'ast str,
        },
        Attribute {
            name: Box<SnapshotAstNode<'ast>>,
            value: SnapshotExpr<'ast>,
        },
        Declaration {
            name: Box<SnapshotAstNode<'ast>>,
            value: SnapshotExpr<'ast>,
            location: Location,
        },
    }

    impl<'ast> SnapshotAstNode<'ast> {
        fn from_node(node: AstNode, content: &'ast str) -> Self {
            match node {
                AstNode::String { value } => {
                    let text = &content[value.to_range()];
                    Self::String { value, text }
                }
                AstNode::Text {
                    value,
                    children,
                    location,
                    attributes,
                } => Self::Text {
                    value: value.map(|n| Box::new(SnapshotAstNode::from_node(*n, content))),
                    text: &content[location.to_range()],
                    attributes: attributes
                        .into_iter()
                        .map(|n| SnapshotAstNode::from_node(n, content))
                        .collect(),
                    children: children
                        .into_iter()
                        .map(|n| SnapshotAstNode::from_node(n, content))
                        .collect(),
                    location,
                },
                AstNode::Span {
                    value,
                    location,
                    attributes,
                } => Self::Span {
                    value: value.map(|n| Box::new(SnapshotAstNode::from_node(*n, content))),
                    attributes: attributes
                        .into_iter()
                        .map(|n| SnapshotAstNode::from_node(n, content))
                        .collect(),
                    text: &content[location.to_range()],
                    location,
                },
                AstNode::Identifier { value } => Self::Identifier {
                    text: &content[value.to_range()],
                    value,
                },
                AstNode::Attribute { name, value } => Self::Attribute {
                    name: Box::new(SnapshotAstNode::from_node(*name, content)),
                    value: SnapshotExpr::from_expr(value, content),
                },
                AstNode::Declaration {
                    name,
                    value,
                    location,
                } => Self::Declaration {
                    name: Box::new(SnapshotAstNode::from_node(*name, content)),
                    value: SnapshotExpr::from_expr(value, content),
                    location,
                },
            }
        }
    }

    fn get_ast(template: &str) -> SnapshotAst<'_> {
        let tokens = Lexer::new(template).collect::<Vec<_>>();
        let tokens = Tokens::new(tokens, template.len());
        let parser = Parser::new(tokens);
        SnapshotAst::from_ast(parser.parse(), template)
    }

    #[test]
    fn test_parser() {
        let template = r#"

text "Hello"
    span "World""#;

        let tokens = Lexer::new(template).collect::<Vec<_>>();
        let tokens = Tokens::new(tokens, template.len());
        let parser = Parser::new(tokens);

        let ast = SnapshotAst::from_ast(parser.parse(), template);

        insta::assert_yaml_snapshot!(ast);
    }

    #[test]
    fn test_parsing_weirdly_formatted_attributes() {
        let template = r#"
text [
    foreground
    :
    "red",
    background

    :


    'green',
] "Hello"
    span [background: "yellow"] "World"
"#;

        insta::assert_yaml_snapshot!(get_ast(template));
    }

    #[test]
    fn test_parsing_attributes() {
        let template = r#"
text [foreground: "red"] "Hello"
    span [background: "yellow"] "World"
"#;

        insta::assert_yaml_snapshot!(get_ast(template));
    }
}
