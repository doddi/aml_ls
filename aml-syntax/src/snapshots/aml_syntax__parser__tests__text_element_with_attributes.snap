---
source: aml-syntax/src/parser.rs
expression: ast
---
nodes:
  - Text:
      values:
        - String:
            value: "\"Hello\""
            location:
              start_byte: 27
              end_byte: 34
      attributes:
        - Attribute:
            name:
              Identifier:
                value: foreground
                location:
                  start_byte: 6
                  end_byte: 16
            value:
              Primitive:
                original: "#ff0000"
                value:
                  Hex:
                    r: 255
                    g: 0
                    b: 0
                location:
                  start_byte: 18
                  end_byte: 25
            location:
              start_byte: 6
              end_byte: 25
            original: "foreground: #ff0000"
      children: []
      text: "text [foreground: #ff0000] \"Hello\""
      location:
        start_byte: 0
        end_byte: 34
      keyword:
        start_byte: 0
        end_byte: 4
variables: {}
scopes:
  - variables: []
    parent: ~
